File: 955-minimum-number-of-k-consecutive-bit-flips/main.go
func minKBitFlips(nums []int, k int) int {
    
    n := len(nums)
    flips, flipsEffect := 0, 0
    isFlipped := make([]int, n)

    for idx := range n {
        
        if idx >= k {
            flipsEffect ^= isFlipped[idx - k]
        }

        if flipsEffect == nums[idx] {

            if idx + k > n {
                return -1
            }

            flips += 1
            flipsEffect ^= 1
            isFlipped[idx] = 1
        }
    }

    return flips
}

File: 955-minimum-number-of-k-consecutive-bit-flips/main.cpp
class Solution {
public:
    int minKBitFlips(vector<int>& nums, int k) {
        
        int n = nums.size();
        int flips = 0;
        int flip_effect = 0;
        std::vector<int> is_flipped(n, 0);

        for (int i = 0; i < n; ++i) {
            if (i >= k) {
                flip_effect ^= is_flipped[i - k];
            }

            if (flip_effect == nums[i]) {
                if (i + k > n) {
                    return -1;
                }
                flips++;
                flip_effect ^= 1;
                is_flipped[i] = 1;
            }
        }

        return flips;

    }   
};
